app/services/cache_manager.rb

```diff
@@ -10,7 +10,39 @@ def self.create_new_permission_cache
   end
 
   def self.create_new_filter_cache
-    Filter.all.each(&:save!)
+    locked_roles = Role.select(&:locked?)
+    origins = locked_roles.map(&:origin)
+    locked_roles.each { |role| role.origin = nil }
+    locked_roles.each { |role| role.modify_locked = true }
+    locked_roles.each(&:save)
+    Filter.all.each do |filter|
+      next if filter.save
+
+      if filter.errors[:permissions] == ['Permissions must be of same resource type']
+        # If the filter was created long time ago, it may have had several
+        # resources of the same type. To fix it, let's split it into different
+        # filters each with its own resource type.
```

Comment:
Yes, I could make this a separate issue and make it in a migration, it's not a problem.